Backporting master_update_table_statistics improvements to 9.4

naisila opened the pull request:
DESCRIPTION: Improves `master_update_table_statistics` and provides distributed deadlock detection

Users who upgrade to the patch release of 9.4 will get the fix via 9.4-2--9.4-3.

onurctirtir commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -207,6 +207,19 @@ SELECT * FROM print_extension_changes();
+ALTER EXTENSION citus UPDATE TO '9.4-3';
+SELECT * FROM print_extension_changes();
+
+-- Snapshot of state at 9.4-2
+ALTER EXTENSION citus UPDATE TO '9.4-2';
+SELECT * FROM print_extension_changes();
+
```
I think we should not have this part:

```suggestion
```

(I guess the reason behind the confusion is that we didn't add a similar test 9.4-2)

SaitTalhaNisanci replied:
We don't have a direct path from 9.4-3 to 9.5. Actually we don't have a direct path from 9.4-2 to 9.5 so maybe we should have 9.4-1 here? To go to 9.5 we will follow 9.4-3 -> 9.4-2 -> 9.4.1 -> 9.5.1

onurctirtir reviewed (approved):
Cherry-pick's for 7db8c85e07a535983308e514169686cde757a475 & c1fa470168fc208e45182cdc38a90138ec78ba27 look good to me.

For the other commits, let's wait for others' input too.

SaitTalhaNisanci commented on src/backend/distributed/sql/citus--9.4-2--9.4-3.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE OR REPLACE FUNCTION master_update_table_statistics(relation regclass)
+RETURNS VOID
+    LANGUAGE C STRICT
+    AS 'MODULE_PATHNAME', $$citus_update_table_statistics$$;
```
I guess we started using the `citus` prefix instead of `master` as of Citus 10, but since this is an internal function, it is probably okay to use `citus` prefix here

naisila replied:
That's right, but `master_update_table_statistics` doesn't exist as an internal function, the internal function `citus_update_table_statistics` was introduced with improvement of this functionality in the commit we originally want to backport https://github.com/citusdata/citus/commit/2f30614fe343103f2c90c804f7c6e4c9d1cbfc71 I wanted to stay consistent with the original commit therefore the choice of using the name `citus_update_table_statistics` for the internal function. As you said, it shouldn't be a problem since it's an internal function.

SaitTalhaNisanci commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -207,6 +207,15 @@ SELECT * FROM print_extension_changes();
+
+ALTER EXTENSION citus UPDATE TO '9.4-3';
+SELECT * FROM print_extension_changes();
+
```
I think we should add 9.4-1 here as the upgrade will go from 9.4-1 to 9.5-1.

hanefi replied:
I do not think that `release-9.4` branch needs to care about `9.5-1`. However your point may be valid on release branches for more recent major versions.

SaitTalhaNisanci commented:
I think we should change the DESCRIPTION line to reflect what this backport fixes. Maybe use the original one: `Improves citus_update_table_statistics and provides distributed deadlock detection`

SaitTalhaNisanci commented:
I have manually verified that I am getting the undetected distributed deadlock before this commit on 9.4 and not after the fix.

naisila commented:
@SaitTalhaNisanci I force pushed but here are the changes to scripts and tests:

- I added `SELECT prosrc FROM pg_proc WHERE proname = 'master_update_table_statistics';` to tests to print the difference of `master_update_table_statistics` when upgrading/downgrading
**addressing this comment** https://github.com/citusdata/citus/pull/5152#discussion_r679963999

- I added `ALTER EXTENSION citus UPDATE TO '9.4-1';` to tests as the main upgrade path will go from 9.4-1 to 9.5-1.
**addressing this comment** https://github.com/citusdata/citus/pull/5147#discussion_r679929157

- I added explanation to the downgrade script
**addressing this comment** https://github.com/citusdata/citus/pull/5152#discussion_r679965860
